File: JavaAbril30dias/src/Dia6.java
import java.util.Scanner;
import java.util.ArrayList;

public class Dia6 {
    public static void main (String[]args) {
        double[] changeTypes = {500, 200, 100, 50, 20, 10, 5, 2, 1, 0.5, 0.2, 0.1, 0.05, 0.02, 0.01};
        Scanner input = new Scanner(System.in);
        int counter = 1;
        String cashType;
        ArrayList<Double> changeReturned = new ArrayList<>();

        System.out.println("¿Con cuanto dinero vas a pagar?:");
        double money = input.nextDouble();
        System.out.println("¿Cuanto cuesta el producto?:");
        double price = input.nextDouble();

        double change = Math.round((money - price * 1.21) * 100.0) / 100.0;

        if (change < 0) {
            System.out.println("Dinero insuficiente");
        }
        else {
            while (change > 0) {
                for (double i : changeTypes) {
                    if (i <= change) {
                        changeReturned.add(i);
                        change -= i;
                        change = Math.round(change * 100.0) / 100.0;
                        break;
                    }
                }
            }

            for (int i=1; i < changeReturned.size(); i++) {
                if (changeReturned.get(i).equals(changeReturned.get(i-1))) {
                    counter += 1;
                }
                else {
                    cashType = changeReturned.get(i) >= 5 ? "billete" : "moneda";
                    if (counter > 1) {
                        cashType += "s";
                    }
                    System.out.println(counter + " " + cashType + " de " + changeReturned.get(i) + " €.");
                    counter = 1;
                }
            }
            cashType = changeReturned.getLast() >= 5 ? "billete" : "moneda";
            if (counter > 1) {
                cashType += "s";
            }
            System.out.println(counter + " " + cashType + " de " + changeReturned.getLast() + " €.");
        }
    }
}

File: PythonAbril30dias/Dia6.py
# Lista de dinero en efectivo
change_types = [500, 200, 100, 50, 20, 10, 5, 2, 1, 0.5, 0.2, 0.1, 0.05, 0.02, 0.01]
# Contador para agrupar las monedas/billetes del mismo tipo
counter = 1
cash_type = None
change_returned = []
money = float(input("¿Con cuanto dinero vas a pagar?: "))
price = float(input("¿Cuanto cuesta el producto?: "))
change = money - round(price*1.21, 2)


if change < 0:
    print("Dinero insuficiente")
else:
    while change > 0:
        for i in change_types:
            if i <= change:
                change_returned.append(i)
                change -= i
                change = round(change, 2)
                break

    for i in range(1, len(change_returned)):
        if change_returned[i] == change_returned[i - 1]:
            counter += 1
        else:
            cash_type = "billete" if change_returned[i - 1] >= 5 else "moneda"
            if counter > 1:
                cash_type = cash_type + "s"
            print(f"{counter} {cash_type} de {change_returned[i - 1]}€.")
            counter = 1

    cash_type = "billete" if change_returned[-1] >= 5 else "moneda"
    if counter > 1:
        cash_type = cash_type + "s"
    print(f"{counter} {cash_type} de {change_returned[-1]}€.")


